### ISSUE TITLE
strace.py -f breaks with system calls's siginal

### ISSUE THREAD

**Issue by IdearHui** [haypo/python-ptrace#35]
when I use "strace.py -f " to tracing ansible apt module, it throw me a bug that system call breaks. I used gdb to debug but couldn't attach on the proess. Using ps to check process I found that python-ptrace was still monitoring the ansible module, So I killed the forks, and the ptrace received an interrupt siginal.
I am curious about why adding the argument "-f " may lead the python-ptrace program getting stuck?

**Comment by haypo** [comment]
It would help if you give more information: what is the traced application? Does the application have threads? Does it spawn child processes? Can you paste the last lines of the file before strace.py hangs?

**Comment by IdearHui** [comment]
I was tracing an ansible command.
the command was:
**strace.py** -f -- /bin/sh -c '/usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1486628823.51-187444217093550/apt.py && sleep 0'
here is the last few lins of the result
![image](https://cloud.githubusercontent.com/assets/14813460/22822184/4c1dafe6-efb9-11e6-8701-20378152aa2e.png)
![image](https://cloud.githubusercontent.com/assets/14813460/22822189/544ca41a-efb9-11e6-8d25-ab453fb64891.png)
Picture2 is the last few lines, as is shown in picture1, we have found that arguments in function fstat is different from nomal result(in picture below), if is the reason caused the following function read receive wrong arguments?
here is the result for command --- **strace** -f /bin/sh -c '/usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1486628823.51-187444217093550/apt.py && sleep 0'
(the trace here is linux shell command)
![image](https://cloud.githubusercontent.com/assets/14813460/22822527/ccf8d162-efba-11e6-9249-6b64979c289a.png)

if you wanna the program apt.py, I can e-mail you

**Comment by trnila** [comment]
I have similar issue with following program
```c
#include <stdio.h>                                                                                                                                             
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <wait.h>

void handler(int sig) {}

int main() {
    int parent = getpid();
    int child = fork();
    if(child == 0) {
        sleep(1);
        kill(parent, SIGUSR1);

        exit(0);
    }
    signal(SIGUSR1, handler);

    pause();

    while(wait(NULL) > 0);
}
```

When I replaced all occurences of *process* with *event.process* it is not freezing anymore. https://github.com/haypo/python-ptrace/blob/master/strace.py#L161-L185

When Exception is thrown it wont set *process* variable (https://github.com/haypo/python-ptrace/blob/master/strace.py#L169)

**Comment by haypo** [comment]
> When I replaced all occurences of process with event.process it is not freezing anymore.

I'm not able to repdroduce the bug with the C program, but I pushed the proposed fix: commit 45c3f0697e0cc72192f18311fd64d639221e62bb.

@IdearHui: Does it fix your bug? If yes, I can release a new version including the fix.

**Comment by IdearHui** [comment]
@haypo Thanks a lot ! The commit [45c3f06](https://github.com/haypo/python-ptrace/commit/45c3f0697e0cc72192f18311fd64d639221e62bb) really helps !

**Comment by haypo** [comment]
I just released python-ptrace 0.9.2 which contains the fix. Thanks Daniel Trnka for the fix, thanks @IdearHui for the bug report.
